packages/ra-ui-materialui/src/list/datagrid/Datagrid.tsx

```diff
@@ -229,60 +230,62 @@ export const Datagrid: FC<DatagridProps> = React.forwardRef((props, ref) => {
      */
     return (
         <DatagridContextProvider value={contextValue}>
-            <DatagridRoot
-                sx={sx}
-                className={clsx(DatagridClasses.root, className)}
-            >
-                {bulkActionButtons !== false ? (
-                    <BulkActionsToolbar>
-                        {isValidElement(bulkActionButtons)
-                            ? bulkActionButtons
-                            : defaultBulkActionButtons}
-                    </BulkActionsToolbar>
-                ) : null}
-                <div className={DatagridClasses.tableWrapper}>
-                    <Table
-                        ref={ref}
-                        className={DatagridClasses.table}
-                        size={size}
-                        {...sanitizeRestProps(rest)}
-                    >
-                        {createOrCloneElement(
-                            header,
-                            {
-                                children,
-                                sort,
-                                data,
-                                hasExpand: !!expand,
-                                hasBulkActions,
-                                isRowSelectable,
-                                onSelect,
-                                resource,
-                                selectedIds,
-                                setSort,
-                            },
-                            children
-                        )}
-                        {createOrCloneElement(
-                            body,
-                            {
-                                expand,
-                                rowClick,
-                                data,
-                                hasBulkActions,
-                                hover,
-                                onToggleItem: handleToggleItem,
-                                resource,
-                                rowSx,
-                                rowStyle,
-                                selectedIds,
-                                isRowSelectable,
-                            },
-                            children
-                        )}
-                    </Table>
-                </div>
-            </DatagridRoot>
+            <OptionalResourceContextProvider value={resource}>
```

Comment:
Stricter types revealed a bug: Using a Datagrid in controlled mode with an EditButton as a child used to fail silently (the EditButton link didn't work), because the resource set on the Datagrid was never propagated to the child. Now the failure is loud ans forces to add this ResourceContext.